src/variety/pencils.js

```diff
@@ -0,0 +1,804 @@
+
+		if(dir>= 1 && dir <= 4) {
+			cell.setPencilArrow(0, true);
+		} else {
+			this.inputclean_cell();
+		}
+	},
+
+	inputarrow_cell_main : function(cell, dir){
+		cell.setPencilArrow(dir, this.puzzle.editmode);
+	},
+
+	inputqnum_pencils : function(){
+		var cell = this.getcell();
+		if(cell.isnull){ return;}
+
+		if(cell!==this.cursor.getc() && this.inputMode==='auto'){
+			this.setcursor(cell);
+		}
+		else{
+			this.inputcell_pencils(cell);
+		}
+	},
+	
+	inputcell_pencils : function(cell){
+		var dir = cell.qdir, num = cell.qnum, val;
+		// -4to-1:Arrow 0:? 1:何もなし 2以上:数字
+		if     (num===-2){ val = 0;}
+		else if(dir=== 5){ val = (num!==-1 ? num+1 : 2);}
+		else if(dir=== 0){ val = 1;}
+		else             { val = dir - 5;}
+
+		var min = -4, max = cell.getmaxnum()+1;
+		if(this.inputMode.match(/number/)){ min = 1;}
+		if(this.inputMode==='undef'){ max = 1; min = 0;}
+
+		if(this.btn==='left'){
+			if(min<=val && val<max){ val++;  }
+			else                   { val=min;}
+		}
+		else if(this.btn==='right'){
+			if(min<val && val<=max){ val--;  }
+			else                   { val=max;}
+		}
+
+		if(val >= 0) { cell.setPencilArrow(0, true);}
+
+		if     (val >=2){ cell.setQdir(5);     cell.setNum(val-1);}
+		else if(val===1){ cell.setQdir(0);     cell.setNum(-1);}
+		else if(val===0){ cell.setQdir(5);     cell.setNum(-2);}
+		else            { cell.setPencilArrow(val+5, true);}
+		cell.draw();
+	}
+},
+
+KeyEvent:{
+	enablemake : true,
+	moveTarget : function(ca){
+		if(ca.match(/shift/)){ return false;}
+		return this.moveTCell(ca);
+	},
+
+	keyinput : function(ca){
+		var cell = this.cursor.getc();
+		var dir = 0;
+		switch(ca){
+			case 'shift+up':    dir = cell.UP; break;
+			case 'shift+down':  dir = cell.DN; break;
+			case 'shift+left':  dir = cell.LT; break;
+			case 'shift+right': dir = cell.RT; break;
+		}
+
+		if(dir !== 0 && !cell.isnull) {
+			cell.setPencilArrow(dir, true);
+			this.cursor.draw();
+		} else {
+			this.key_inputqnum_pencils(ca);
+		}
+	},
+
+	key_inputqnum_pencils : function(ca){
+		var cell = this.cursor.getc();
+		if(ca==='q' || ca==='-'){
+			if(cell.qnum !== -2) {
+				cell.setPencilArrow(0, true);
+				cell.setQdir(5);
+				cell.setQnum(-2);
+			} else {
+				cell.setPencilArrow(0, true);
+				cell.setQnum(-1);
+			}
+		}
+		else if(ca==='BS' && cell.qdir===5){
+			if(cell.qnum!==-1){
+				this.key_inputqnum_main(cell,ca);
+				if(cell.qnum===-2){
+					cell.setQnum(-1);
+				}
+			}
+		}
+		else if(ca===' ' || ca==='BS'){
+			cell.setPencilArrow(0, true);
+			cell.setQnum(-1);
+		}
+		else{
+			this.key_inputqnum_main(cell,ca);
+			if(cell.isNum() && cell.qdir!==5) {
+				cell.setPencilArrow(0, true);
+				cell.setQdir(5);
+			}
+		}
+
+		this.prev = cell;
+		cell.draw();
+	}
+},
+
+//---------------------------------------------------------
+// 盤面管理系
+Cell:{
+	maxnum : function(){
+		var bd=this.board;
+		return Math.max(bd.cols,bd.rows)-1;
+	},
+	minnum : 1,
+
+	setPencilArrow: function(dir, question) {
```

Comment:
somehow I wish this whole method were a bit more concise, not sure it's a worthwhile change though...

maybe helpers like`this.getAdjBorder(dir)`, `this.setArrow(dir, dirborder, question)`, `this.clearArrow(dir, dirborder, question)`?

Reply:
I've rewritten this function, so it feels a bit less copy-and-paste.

Reply:
great, thank you!